crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -0,0 +1,720 @@
+    //When an instruction performs arithmetic operation, its output can be represented as an arithmetic expression of its arguments
+    //Substitute a nodeobj as an arithmetic expression
+    fn substitute(
+        arith_cache: &mut HashMap<Index, InternalVar>,
+        idx: Index,
+        evaluator: &mut Evaluator,
+        cfg: &IRGenerator,
+    ) -> InternalVar {
+        if arith_cache.contains_key(&idx) {
+            return arith_cache[&idx].clone();
+        }
+        let mut expr;
+        let mut result: Option<InternalVar> = None;
+
+        if let Some(obj) = cfg.get_object(idx) {
+            match obj {
+                node::NodeObj::Const(c) => {
+                    let f_value = FieldElement::from_be_bytes_reduce(&c.value.to_bytes_be()); //TODO const should be a field
+                    expr = Arithmetic {
+                        mul_terms: Vec::new(),
+                        linear_combinations: Vec::new(),
+                        q_c: f_value, //TODO handle other types
+                    };
+                    result = Some(InternalVar {
+                        idx: Some(idx),
+                        expression: expr,
+                        witness: None,
+                    });
+                }
+                node::NodeObj::Obj(v) => {
+                    let w = if let Some(w1) = v.witness {
+                        w1
+                    } else {
+                        evaluator.add_witness_to_cs()
+                    };
+                    expr = Arithmetic::from(&w);
+                    result = Some(InternalVar {
+                        idx: Some(idx),
+                        expression: expr,
+                        witness: Some(w),
+                    });
+                }
+                _ => (),
+            }
+        }
+        if result.is_none() {
+            let w = evaluator.add_witness_to_cs();
+            expr = Arithmetic::from(&w);
+            result = Some(InternalVar {
+                idx: Some(idx),
+                expression: expr,
+                witness: Some(w),
+            });
+        }
+        arith_cache.insert(idx, result.unwrap());
+        arith_cache[&idx].clone()
+    }
+
+    pub fn new() -> Acir {
+        Acir {
+            arith_cache: HashMap::new(),
+        }
+    }
+
+    pub fn evaluate_instruction(
+        &mut self,
+        ins: &Instruction,
+        evaluator: &mut Evaluator,
+        cfg: &IRGenerator,
+    ) {
+        if ins.operator == Operation::nop {
+            return;
+        }
+        let mut output = Arithmetic::default();
+        let l_c = Acir::substitute(&mut self.arith_cache, ins.lhs, evaluator, cfg);
+        let r_c = Acir::substitute(&mut self.arith_cache, ins.rhs, evaluator, cfg);
+        match ins.operator {
+            Operation::add | Operation::sadd => {
+                //output = &l_c.expression + &r_c.expression;
+                output = add(&l_c.expression, FieldElement::one(), &r_c.expression);
```

Comment:
The sub and eq_gate need to do stuff with the output so they do not end with output=

Reply:
You can use a `let` for those cases e.g.
```rs
Operation::sub => {
    ...
    let f = FieldElement::from(k);
    let f = f * FieldElement::from_be_bytes_reduce(&BigUint::from(r_mod).to_bytes_be());
    let mut output = add(&l_c.expression, FieldElement::from(-1), &r_c.expression);
    output.q_c += f;
    output
}
```

Reply:
I am sorry I was a bit harsh indeed, I am always glad to have my code looking more rusty, so I thank you for your review and your comments!
I did implement your suggestion.

Reply:
Thanks! No need to apologize though, I don't think you were overly harsh. While we're on the topic though, if you ever think I am overly harsh, feel free to reply to me either here or on slack and tell me so. It can be hard to balance comments sometimes when I want to comment more for a larger PR but commenting too much can feel like I'm overwhelming the other person with suggestions.